File: lua/core/autocmds.lua
local M = {}

function M.preset()
  return {
    ---Hightlight yanked text.
    hl_yank = {
      highgroup = "Search",
      timeout = 200,
    },
    ---Format on save.
    format_on_save = {
      timeout = nil,
    },
    ---Map `q` to close window.
    q_to_close = {
      "vim",
      "help",
      "man",
      "git",
      "qf",
      "lspinfo",
      "TelescopePrompt",
      "null-ls-info",
    },
    ---Remove trailing spaces.
    trim_spaces = true,
    ---Auto resize windows when changed.
    auto_resize = true,
  }
end

---@param minv MiNV
function M.setup(minv)
  local preset = minv.autocmds
  local au = require("utils").autocmd
  local function au_if(enable, ...)
    if enable then
      au(...)
    end
  end

  if preset.hl_yank ~= nil then
    local cmd = string.format(
      [[lua require("vim.highlight").on_yank({higroup = "%s", timeout = %d})]],
      preset.hl_yank.highgroup,
      preset.hl_yank.timeout
    )
    au("TextYankPost", "*", cmd)
  end
  if preset.format_on_save ~= nil then
    local cmd = string.format(
      [[:silent lua vim.lsp.buf.formatting_sync(nil, %s)]],
      preset.format_on_save.timeout
    )
    au("BufWritePre", "*", cmd)
  end
  if preset.q_to_close ~= nil then
    local ft = table.concat(preset.q_to_close, ",")
    au("FileType", ft, "nnoremap <silent> <buffer> q <Cmd>close!<CR>")
  end
  au_if(preset.trim_spaces, "BufWritePre", "*", [[silent s/\s\+$//e]])
  au_if(preset.auto_resize, "VimResized", "*", "wincmd =")
end

return M

File: src/minv/autocmds.ts
import { autocmd } from "./utils";

export type Autocmds = typeof AUTOCMDS;
export const AUTOCMDS = {
  hl_yank: {
    enable: true,
    pattern: "*",
    highgroup: "Search",
    timeout: 200,
  },
  format_on_save: {
    enable: true,
    pattern: "*",
    timeout: undefined,
  },
  close: {
    enable: true,
    key: "q",
    filetype: [
      "vim",
      "help",
      "man",
      "git",
      "qf",
      "lspinfo",
      "TelescopePrompt",
      "null-ls-info",
    ],
  },
  trim_spaces: {
    enable: true,
    pattern: "*",
  },
  auto_resize: {
    enable: true,
    pattern: "*",
  },
  ruler: {
    enable: true,
    offsets: {} as Record<string, number>,
  },
};

// TODO: add descriptions
const ARGS: {
  [K in keyof Autocmds]: (this: void, opts: Autocmds[K]) => void;
} = {
  hl_yank: ({ pattern, highgroup, timeout }) =>
    autocmd(
      "TextYankPost",
      () => {
        (require("vim.highlight") as AnyTbl).on_yank({ highgroup, timeout });
      },
      { pattern }
    ),
  format_on_save: ({ pattern, timeout }) =>
    autocmd(
      "BufWritePre",
      () => {
        vim.lsp.buf.formatting_sync(undefined, timeout);
      },
      { pattern }
    ),
  close: ({ key, filetype }) =>
    autocmd(
      "FileType",
      ({ buf }) => {
        vim.api.nvim_buf_set_keymap(buf, "", key, "<Cmd>q!<CR>", {
          silent: true,
          noremap: true,
        });
      },
      { pattern: filetype }
    ),
  trim_spaces: ({ pattern }) =>
    autocmd("BufWritePre", `silent s/\s+$//e`, { pattern }),
  auto_resize: ({ pattern }) => autocmd("VimResized", "wincmd =", { pattern }),
  ruler: ({ offsets }) => {
    for (const [ft, offs] of pairs(offsets)) {
      autocmd("FileType", string.format("setlocal colorcolumn=%d", offs), {
        pattern: ft,
      });
    }
  },
};

export function setup_autocmds(this: void) {
  for (const [name, opts] of pairs(AUTOCMDS)) {
    if (opts.enable) {
      ARGS[name](opts as any);
    }
  }
}
